pkg/hubble/container/ring_reader.go

```diff
@@ -77,26 +85,50 @@ func (r *RingReader) Next() (*v1.Event, error) {
 func (r *RingReader) NextFollow(ctx context.Context) *v1.Event {
 	// if the context changed between invocations, we also have to restart
 	// readFrom, as the old readFrom instance will be using the old context.
-	if r.c == nil || r.ctx != ctx {
-		r.c = r.ring.readFrom(ctx, r.idx)
+	if r.ctx != ctx {
+		if r.followChan == nil {
+			r.followChan = make(chan *v1.Event, r.followChanLen)
+		}
+		r.wg.Add(1)
+		go func(ctx context.Context) {
+			r.ring.readFrom(ctx, r.idx, r.followChan)
+			if ctx.Err() != nil && r.followChan != nil { // context is done
+				close(r.followChan)
+				r.followChan = nil
+			}
+			r.wg.Done()
+		}(ctx)
 		r.ctx = ctx
 	}
+	defer func() {
+		if ctx.Err() != nil { // context is done
+			r.ctx = nil
+		}
+	}()
 
 	select {
-	case e, ok := <-r.c:
+	case e, ok := <-r.followChan:
 		if !ok {
-			// channel can only be closed by readFrom if ctx is cancelled
-			r.c = nil
-			r.ctx = nil
+			// the channel is closed so the context is done
 			return nil
 		}
 		// increment idx so that future calls to the ring reader will
 		// continue reading from were we stopped.
 		r.idx++
 		return e
 	case <-ctx.Done():
-		r.c = nil
-		r.ctx = nil
 		return nil
 	}
 }
+
+// Close waits for any method to return and closes the RingReader.
+// Callers are not required to call Close on a RingReader but it may be useful
+// for specific situations such as testing.
+func (r *RingReader) Close() error {
+	r.wg.Wait()
+	if r.followChan != nil {
+		close(r.followChan)
+		r.followChan = nil
+	}
```

Comment:
I think this is now not needed anymore. With the closure of the channel on line 94, there is no case where `r.wg.Wait()` unblocks without the channel already being closed, right?